### PR TITLE
special case for __typename in authorization

### PR DESCRIPTION
**Pull request by Geal**
When evaluating authorization directives on fields returning interfaces,
we require the usage of fragments with type conditions if the interface
implementors have different security requirements. For the __typename
field though, we must make an exception, because it should be available
for all implementors

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: .changesets/fix_geal_authorization_typename.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+### special case for __typename in authorization ([PR #3821](https://github.com/apollographql/router/pull/3821))
+
+When evaluating authorization directives on fields returning interfaces,
+we require the usage of fragments with type conditions if the interface
+implementors have different security requirements. For the __typename
+field though, we must make an exception, because it should be available
+for all implementors
+
+<!-- start metadata -->
+---
```
Comment (bnjjj) [inline .changesets/fix_geal_authorization_typename.md:10]:
```suggestion
```

this bug should be fixed now

**Review by bnjjj — APPROVED** [review]

File: .changesets/fix_geal_authorization_typename.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+### special case for __typename in authorization ([PR #3821](https://github.com/apollographql/router/pull/3821))
+
+When evaluating authorization directives on fields returning interfaces,
```
Comment (BrynCooke) [inline .changesets/fix_geal_authorization_typename.md:3]: I don't fully understand this fix. Is there a ticket that goes into more detail?

Could `__typename` represent a leakage of data?

Comment (BrynCooke) [reply .changesets/fix_geal_authorization_typename.md:3]: If a concrete type is not accessible then a caller can still check to see if something exists by asking for the typename.

In the examples below a caller can see that the blog concrete type is a `PrivateBlog` even if they don't have access.

Comment (Geal) [reply .changesets/fix_geal_authorization_typename.md:3]: Yes, it is expected. The authorization feature assumes that clients know about the schema, and the schema indicates the condition to access types or fields. If the data is not there and the subgraph did not return a "not found" explicitely, it's easy to infer that the type was `PrivateBlog`. If we actually want to return "not found" on unauthorized data, it is better handled at the subgraph level

**Review by BrynCooke — APPROVED** [review]

**Pull request merged**